gbra/attackers/attacker.py

```diff
@@ -194,3 +194,114 @@ def attack(self, verbose = False):
+    fake user.
+    """
+
+    def __init__(self, _num_items_to_scout, _num_recs, *args, **kwargs):
+        super(BlackBoxRWRAttacker, self).__init__(*args, **kwargs)
+        self.num_items_to_scout = _num_items_to_scout
+        self.num_recs = _num_recs
+
+    def attack(self, verbose = False):
+        item_to_approx_rwr = self.run_scout()
+
+        max_rating = self.recommender._G.max_rating
+        sorted_items = sorted(
+            item_to_approx_rwr.iteritems(),
+            key=lambda (iid, approx_rwr): approx_rwr, reverse=True
+        )
+
+        for i in range(self.num_fake_entities):
+            fake_entity_id = self.add_fake_entity()
+            item_to_attack = sorted_items[i % len(sorted_items)][0]
+            self.recommender._attacker_add_edge(
+                fake_entity_id, item_to_attack, max_rating
+            )
+            self.recommender._attacker_add_edge(
+                fake_entity_id, self.target_item, max_rating
+            )
+
+    def run_scout(self):
+        scout_id = self.add_fake_entity()
+
+        item_to_approx_rwr = dict(
+            (item, self.approx_rwr(scout_id, item))
+            for item, _ in self.get_top_items()
+        )
+        return item_to_approx_rwr
+
+    def get_top_items(self):
+        """Return the top `num_items_to_scout` according to weighted degree
+
+        Weighted degree for an item is the sum of weights of its edges.
+
+        We originally considered shifting the weights to be centered at 0,
+        with the intuition that an item with many negative reviews should
+        have a lower score than one with few, but positive, reviews.
+
+        However -- we decided to stick with just vanilla weighted degree,
+        because it's not clear that such a weighting scheme leads to items
+        with higher RWR.
+        """
+        return sorted(
+            self.recommender._G.get_weighted_item_to_degree().iteritems(),
+            key=lambda (iid, weighted_deg): weighted_deg,
+            reverse=True
+        )[:self.num_items_to_scout]
+
+    def approx_rwr(self, entity_id, item_id):
+        max_rating = self.recommender._G.max_rating
+        self.recommender._attacker_add_edge(entity_id, item_id, max_rating)
+        recs = self.recommender.recommend(entity_id, self.num_recs)
+        self.recommender._G.del_edge(entity_id, item_id)
+
+        return sum(
+            self.recommender._G.get_weighted_degree(iid)
+            for iid in recs
+        )
+
+class BlackBoxDeepRWRAttacker(BlackBoxRWRAttacker):
+    """A BlackBoxRWRAttacker that goes deeper.
+
+    BlackBoxRWRAttacker approximates RWR by going 1 level deep in the
+    recommendation tree, see approx_rwr().
+
+    BlackBoxDeepRWRAttacker is the same, except that approx_rwr goes
+    '_rec_tree_depth' levels deep when summing the weighted degree of
+    neighbors.
+
+    TODO: allow a variant of this class that only counts the
```

Comment:
I think its OK if we don't get to this. I think we have enough content for today's report